### PR TITLE
AK: Variant<Ts...> / LibWasm: Start implementing a basic WebAssembly binary format parser

### PR DESCRIPTION
**Pull request by alimpfard**
Current state:
- Can parse anything I threw at it (as generated by clang, AssemblyScript, and Rust, also whatever `jq.wasm.wasm` was made with :P)
- Implements most of module instantiation, is missing some data initialisations (passive data will always be zero), tables will not be initialised
- Does _not_ implement validation (this can be done in a future PR)
- Implements most of the instructions, except the following 
<details> <summary>Expand for unimplemented instructions</summary>

- br.table
- table.init
- table.get
- table.set
- table.copy
- table.size
- table.grow
- table.fill
- ref.null
- ref.func
- ref.is_null
- i32/i64.clz
- i32/i64.ctz
- i32/i64.popcnt
- i32/i64.rotl
- i32/i64.rotr
- X.trunc.Y
- X.trunc_sat.Y
- memory.init
- memory.copy
- memory.fill
- elem.drop
- data.drop

</details>

- No JS integration (the WebIDL definitions seemed a bit impossible for our idl wrapper generator)
- `wasm` utility can parse, print, instantiate and execute a function from a wasm module (with user-supplied args)

On the non-Wasm changes (`AK::Variant<Ts...>`)
- Can't be valueless (that's a really bad API)
- Nice `.visit()` method with builtin overload resolution (can pass multiple visitors)
- `.get<T>()` method, will VERIFY that it contains the type requested, use `.get_pointer<T>()` if you're not sure that the type is there, will either return a pointer to the value, or a nullptr

<details>
<summary> 
(Original PR message)
</summary>
This can currently parse and dump a simple module.
This also adds a 'wasm' test utility that tries to parse a given webassembly binary file, and can print out its contents (mostly, instruction arguments are not printed)

image_2021-04-27_22-06-23
</details>

### CODE REVIEW

File: Userland/Libraries/LibWasm/Parser/LEB128.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+#pragma once
+
+#include <AK/Stream.h>
+#include <AK/Types.h>
+
+namespace Wasm {
+
+struct LEB128 {
```
Comment (awesomekling) [inline Userland/Libraries/LibWasm/Parser/LEB128.h:14]: There's LEB128 code in LibDebug already, can we find a way to share this?

Comment (alimpfard) [reply Userland/Libraries/LibWasm/Parser/LEB128.h:14]: I assume you mean the one in `AK/MemoryStream.h`?
This is actually a copy of that, but that one wants to un-read whatever it has read when it fails.
I neither need nor want that behaviour here.

We could implement it in `InputStream`, but that seems like a really odd thing to add to a stream.
One thing I could do is to move it to `AK/LEB128.h` and make both impls use that

Comment (awesomekling) [reply Userland/Libraries/LibWasm/Parser/LEB128.h:14]: >One thing I could do is to move it to `AK/LEB128.h` and make both impls use that

👍🏻

**Comment by alimpfard** [comment]
With _these_ changes, now we can parse every damn single instruction in wasm.
Here's a fun trace, parsing this C function:
```c
int foo(int x, int y) {
    if (x > 4)
        return x + y;
    if (x > y)
        return x + y;
    else
        return 42;
}
```

<details>
<summary>Expand for fun</summary>

```
entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:1134] (Module)
 entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:759] (TypeSection)
  entering [parse_vector<Wasm::FunctionType> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
   entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:165] (FunctionType)
    entering [parse_vector<Wasm::ValueType> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
     entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:127] (ValueType)
     leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:127] (ValueType)
     entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:127] (ValueType)
     leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:127] (ValueType)
    leaving [parse_vector<Wasm::ValueType> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
    entering [parse_vector<Wasm::ValueType> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
     entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:127] (ValueType)
     leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:127] (ValueType)
    leaving [parse_vector<Wasm::ValueType> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
   leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:165] (FunctionType)
  leaving [parse_vector<Wasm::FunctionType> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
 leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:759] (TypeSection)
 entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:804] (ImportSection)
  entering [parse_vector<Wasm::ImportSection::Import> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
   entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:770] (Import)
    entering [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
     entering [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
     leaving [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
    leaving [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
    entering [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
     entering [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
     leaving [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
    leaving [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
    entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:218] (MemoryType)
     entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:190] (Limits)
     leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:190] (Limits)
    leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:218] (MemoryType)
   leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:770] (Import)
   entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:770] (Import)
    entering [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
     entering [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
     leaving [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
    leaving [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
    entering [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
     entering [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
     leaving [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
    leaving [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
    entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:229] (TableType)
     entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:127] (ValueType)
     leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:127] (ValueType)
     entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:190] (Limits)
     leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:190] (Limits)
    leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:229] (TableType)
   leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:770] (Import)
  leaving [parse_vector<Wasm::ImportSection::Import> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
 leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:804] (ImportSection)
 entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:815] (FunctionSection)
  entering [parse_vector<long unsigned int> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
  leaving [parse_vector<long unsigned int> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
 leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:815] (FunctionSection)
 entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:1059] (CodeSection)
  entering [parse_vector<Wasm::CodeSection::Code> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
   entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:1036] (Code)
    entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:1022] (Func)
     entering [parse_vector<Wasm::Locals> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
     leaving [parse_vector<Wasm::Locals> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
     entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:876] (Expression)
      entering [parse_until_any_of<Wasm::Instruction, int> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:94]
       entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
        entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:265] (BlockType)
        leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:265] (BlockType)
        entering [parse_until_any_of<Wasm::Instruction, int, int> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:94]
         entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
         leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
        leaving [parse_until_any_of<Wasm::Instruction, int, int> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:94]
       leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
       leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:296] (Instruction)
      leaving [parse_until_any_of<Wasm::Instruction, int> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:94]
     leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:876] (Expression)
    leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:1022] (Func)
   leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:1036] (Code)
  leaving [parse_vector<Wasm::CodeSection::Code> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
 leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:1059] (CodeSection)
 entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:738] (CustomSection)
  entering [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
   entering [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
   leaving [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
  leaving [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
 leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:738] (CustomSection)
 entering [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:738] (CustomSection)
  entering [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
   entering [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
   leaving [parse_vector<unsigned char> @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:29]
  leaving [parse_name @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:76]
 leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:738] (CustomSection)
leaving [parse @ ../../Userland/Libraries/LibWasm/Parser/Parser.cpp:1134] (Module)
```

</details>

**Comment by alimpfard** [comment]
With a stubbed out `interpret()`, it can now "run" things too
```
Executing - Function with address 0, ptr = 0x01ed0598
    wasm function? true
    type:
      (type function
        (parameters
          (type i32)
          (type i32)
        
        (results
          (type i32)
        )
      )
    code:
      (function
        (locals
          (type i32)
          (type i32)
          (type i32)
        )
        (body
          (i32.const...)
          (local.set...)
          (i32.const...)
          (local.set...)
          (i32.const...)
          (local.set...)
          (block...)
          (local.get...)
        )
      )
wasm(28): FIXME: Interpret stuff!
-> 0 (type i32)
```

**Comment by alimpfard** [comment]
```c
int foo() {
    int x = 42;
    return x;
}
```

image_2021-05-01_03-24-15

:^)

**Comment by alimpfard** [comment]
`fib()`
image_2021-05-03_11-33-24

Timing is rather crap, but that's to be expected :P

**Comment by alimpfard** [comment]
Closing this to split it up into multiple bits, as @bgianfo suggested.

**Pull request closed without merging**